### ISSUE TITLE
[R] [GPU] Unable to load xgboost.so during install

### ISSUE THREAD

**Issue by SteveBronder** [dmlc/xgboost#2789]
## Summary

After compilation of the .so with GPU support for R, the installation in the R session fails because of an undefined symbol
```
Error in dyn.load(file, DLLpath = DLLpath, ...) : 
  unable to load shared object '/home/steve/R/x86_64-pc-linux-gnu-library/3.3/xgboost/libs/xgboost.so':
  /home/steve/R/x86_64-pc-linux-gnu-library/3.3/xgboost/libs/xgboost.so: undefined symbol: 
_ZN7xgboost9Predictor4InitERKSt6vectorISt4pairISsSsESaIS3_EERKS1_ISt10shared_ptrINS_7DMatrixEESaISA_EE
```

## Environment info
Operating System: Ubuntu 17.04 64bit

Compiler: 
In config.mk and R
g++/gcc:
```
$ g++ -v
Using built-in specs.
COLLECT_GCC=g++
COLLECT_LTO_WRAPPER=/usr/lib/gcc/x86_64-linux-gnu/6/lto-wrapper
Target: x86_64-linux-gnu
Configured with: ../src/configure -v --with-pkgversion='Ubuntu 6.3.0-12ubuntu2' --with-bugurl=file:///usr/share/doc/gcc-6/README.Bugs --enable-languages=c,ada,c++,java,go,d,fortran,objc,obj-c++ --prefix=/usr --program-suffix=-6 --program-prefix=x86_64-linux-gnu- --enable-shared --enable-linker-build-id --libexecdir=/usr/lib --without-included-gettext --enable-threads=posix --libdir=/usr/lib --enable-nls --with-sysroot=/ --enable-clocale=gnu --enable-libstdcxx-debug --enable-libstdcxx-time=yes --with-default-libstdcxx-abi=new --enable-gnu-unique-object --disable-vtable-verify --enable-libmpx --enable-plugin --enable-default-pie --with-system-zlib --disable-browser-plugin --enable-java-awt=gtk --enable-gtk-cairo --with-java-home=/usr/lib/jvm/java-1.5.0-gcj-6-amd64/jre --enable-java-home --with-jvm-root-dir=/usr/lib/jvm/java-1.5.0-gcj-6-amd64 --with-jvm-jar-dir=/usr/lib/jvm-exports/java-1.5.0-gcj-6-amd64 --with-arch-directory=amd64 --with-ecj-jar=/usr/share/java/eclipse-ecj.jar --with-target-system-zlib --enable-objc-gc=auto --enable-multiarch --disable-werror --with-arch-32=i686 --with-abi=m64 --with-multilib-list=m32,m64,mx32 --enable-multilib --with-tune=generic --enable-checking=release --build=x86_64-linux-gnu --host=x86_64-linux-gnu --target=x86_64-linux-gnu
Thread model: posix
gcc version 6.3.0 20170406 (Ubuntu 6.3.0-12ubuntu2)
```

for nvcc flag `-DCUDA_HOST_COMPILER=/usr/bin/gcc-4.9`
```
$ gcc-4.9 -v
Using built-in specs.
COLLECT_GCC=gcc-4.9
COLLECT_LTO_WRAPPER=/usr/lib/gcc/x86_64-linux-gnu/4.9/lto-wrapper
Target: x86_64-linux-gnu
Configured with: ../src/configure -v --with-pkgversion='Ubuntu 4.9.4-2ubuntu1' --with-bugurl=file:///usr/share/doc/gcc-4.9/README.Bugs --enable-languages=c,c++,java,go,d,fortran,objc,obj-c++ --prefix=/usr --program-suffix=-4.9 --enable-shared --enable-linker-build-id --libexecdir=/usr/lib --without-included-gettext --enable-threads=posix --with-gxx-include-dir=/usr/include/c++/4.9 --libdir=/usr/lib --enable-nls --with-sysroot=/ --enable-clocale=gnu --enable-libstdcxx-debug --enable-libstdcxx-time=yes --enable-gnu-unique-object --disable-vtable-verify --enable-plugin --with-system-zlib --disable-browser-plugin --enable-java-awt=gtk --enable-gtk-cairo --with-java-home=/usr/lib/jvm/java-1.5.0-gcj-4.9-amd64/jre --enable-java-home --with-jvm-root-dir=/usr/lib/jvm/java-1.5.0-gcj-4.9-amd64 --with-jvm-jar-dir=/usr/lib/jvm-exports/java-1.5.0-gcj-4.9-amd64 --with-arch-directory=amd64 --with-ecj-jar=/usr/share/java/eclipse-ecj.jar --enable-objc-gc --enable-multiarch --disable-werror --with-arch-32=i686 --with-abi=m64 --with-multilib-list=m32,m64,mx32 --enable-multilib --with-tune=generic --enable-checking=release --build=x86_64-linux-gnu --host=x86_64-linux-gnu --target=x86_64-linux-gnu
Thread model: posix
gcc version 4.9.4 (Ubuntu 4.9.4-2ubuntu1) 
```

nvcc:
```
$ nvcc --version
nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2016 NVIDIA Corporation
Built on Sun_Sep__4_22:14:01_CDT_2016
Cuda compilation tools, release 8.0, V8.0.44
```

Package used (python/R/jvm/C++): R

`xgboost` version used: Source

If installing from source, please provide 

1. The commit hash (`git rev-parse HEAD`)
78c4188cec31425f708d238160ea3afb67a7250a

2. Logs will be helpful (If logs are large, please upload as attachment).
See full log here:
https://gist.github.com/SteveBronder/86c17f3ad4dddf7f2ffe86cb2e7203af

TL;DR
```
> deps = setdiff(c('statar','data.table', 'magrittr', 'stringi'), rownames(installed.packages()));    if(length(deps)>0) install.packages(deps, repo = 'https://cloud.r-project.org/')
> 
> 
* installing to library ‘/home/steve/R/x86_64-pc-linux-gnu-library/3.4’
* installing *source* package ‘xgboost’ ...
** libs
installing to /home/steve/R/x86_64-pc-linux-gnu-library/3.4/xgboost/libs
** R
** data
** demo
** preparing package for lazy loading
** help
*** installing help indices
** building package indices
Warning in readLines(file) :
  incomplete final line found on './demo/00Index'
** installing vignettes
** testing if installed package can be loaded
Error: package or namespace load failed for ‘xgboost’ in dyn.load(file, DLLpath = DLLpath, ...):
 unable to load shared object '/home/steve/R/x86_64-pc-linux-gnu-library/3.4/xgboost/libs/xgboost.so':
  /home/steve/R/x86_64-pc-linux-gnu-library/3.4/xgboost/libs/xgboost.so: undefined symbol: _ZN7xgboost9Predictor4InitERKSt6vectorISt4pairISsSsESaIS3_EERKS1_ISt10shared_ptrINS_7DMatrixEESaISA_EE
Error: loading failed
Execution halted
ERROR: loading failed
* removing ‘/home/steve/R/x86_64-pc-linux-gnu-library/3.4/xgboost’
```
If you are using R package, please provide

1. The R `sessionInfo()`
```
> sessionInfo()
R version 3.4.2 (2017-09-28)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 17.04

Matrix products: default
BLAS: /usr/lib/libblas/libblas.so.3.7.0
LAPACK: /usr/lib/lapack/liblapack.so.3.7.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

loaded via a namespace (and not attached):
[1] compiler_3.4.2
```
## Steps to reproduce

2. The command to install `xgboost` if you are not installing from source

```
git clone --recursive https://github.com/dmlc/xgboost
cd xgboost
mkdir build
cd build
cmake .. -DUSE_CUDA=ON -DR_LIB=ON -DCUDA_HOST_COMPILER=/usr/bin/gcc-4.9 -DPLUGIN_UPDATER_GPU=ON
make install -j
```
## What have you tried?

1. Googling / checking past issues
2. Compiling with other / same versions of the compilers
3. Not sure what causes this sort of thing

**Comment by khotilov** [comment]
Could you post the output of `ldd build/R-package/src/xgboost.so`?

**Comment by SteveBronder** [comment]
```
$ ldd build/R-package/src/xgboost.so
	linux-vdso.so.1 =>  (0x00007fff0df04000)
	libR.so => /usr/lib/libR.so (0x00007f2407096000)
	libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f2406e78000)
	libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007f2406c74000)
	librt.so.1 => /lib/x86_64-linux-gnu/librt.so.1 (0x00007f2406a6c000)
	libstdc++.so.6 => /usr/lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007f24066e4000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f24063d9000)
	libgomp.so.1 => /usr/lib/x86_64-linux-gnu/libgomp.so.1 (0x00007f24061ab000)
	libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007f2405f94000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f2405bcd000)
	/lib64/ld-linux-x86-64.so.2 (0x000055d1385e1000)
	libblas.so.3 => /usr/lib/libblas.so.3 (0x00007f2405960000)
	libreadline.so.7 => /lib/x86_64-linux-gnu/libreadline.so.7 (0x00007f2405716000)
	libpcre.so.3 => /lib/x86_64-linux-gnu/libpcre.so.3 (0x00007f24054a1000)
	liblzma.so.5 => /lib/x86_64-linux-gnu/liblzma.so.5 (0x00007f240527b000)
	libbz2.so.1.0 => /lib/x86_64-linux-gnu/libbz2.so.1.0 (0x00007f240506b000)
	libz.so.1 => /lib/x86_64-linux-gnu/libz.so.1 (0x00007f2404e4f000)
	libicuuc.so.57 => /usr/lib/x86_64-linux-gnu/libicuuc.so.57 (0x00007f2404aa7000)
	libicui18n.so.57 => /usr/lib/x86_64-linux-gnu/libicui18n.so.57 (0x00007f240462d000)
	libtinfo.so.5 => /lib/x86_64-linux-gnu/libtinfo.so.5 (0x00007f2404402000)
	libicudata.so.57 => /usr/lib/x86_64-linux-gnu/libicudata.so.57 (0x00007f2402985000)
```

**Comment by khotilov** [comment]
Hmm... I don't see anything alarming. 
Does building and installation without GPU support work?
```
cmake .. -DR_LIB=ON
make install -j
```

I am not sure how big of a hassle these days is to run Ubuntu 17 in Docker or VM with CUDA support, but I'll look into it.

**Comment by SteveBronder** [comment]
Apologies for the delay in my response, building and installation without GPU support does work

**Comment by davidlamcm** [comment]
I got the exact same problem trying to install xgboost with GPU support on ubuntu

`
* installing *source* package ‘matrixStats’ ...
** package ‘matrixStats’ successfully unpacked and MD5 sums checked
** libs
gcc -shared -L/usr/lib/R/lib -Wl,-Bsymbolic-functions -Wl,-z,relro -o matrixStats.so 000.init.o allocMatrix2.o anyMissing.o binCounts.o binMeans.o colCounts.o colOrderStats.o colRanges.o diff2.o indexByRow.o logSumExp.o mean2.o productExpSumLog.o psortKM.o rowCounts.o rowCummaxs.o rowCummins.o rowCumprods.o rowCumsums.o rowDiffs.o rowLogSumExp.o rowMads.o rowMeans2.o rowMedians.o rowOrderStats.o rowRanges.o rowRanksWithTies.o rowSums2.o rowVars.o signTabulate.o sum2.o validateIndices.o weightedMean.o weightedMedian.o x_OP_y.o -L/usr/lib/R/lib -lR
installing to /home/david/R/x86_64-pc-linux-gnu-library/3.3/matrixStats/libs
** R
** inst
** byte-compile and prepare package for lazy loading
** help
*** installing help indices
** building package indices
** installing vignettes
** testing if installed package can be loaded
* DONE (matrixStats)
* installing *source* package ‘statar’ ...
** package ‘statar’ successfully unpacked and MD5 sums checked
** R
** inst
** preparing package for lazy loading
** help
*** installing help indices
** building package indices
** installing vignettes
** testing if installed package can be loaded
* DONE (statar)

The downloaded source packages are in
	‘/tmp/Rtmp8Rv6y4/downloaded_packages’
> 
> 
* installing to library ‘/home/david/R/x86_64-pc-linux-gnu-library/3.3’
* installing *source* package ‘xgboost’ ...
** libs
installing to /home/david/R/x86_64-pc-linux-gnu-library/3.3/xgboost/libs
** R
** data
** demo
** preparing package for lazy loading
** help
*** installing help indices
** building package indices
Warning in readLines(file) :
  incomplete final line found on './demo/00Index'
** installing vignettes
** testing if installed package can be loaded
Error in dyn.load(file, DLLpath = DLLpath, ...) : 
  unable to load shared object '/home/david/R/x86_64-pc-linux-gnu-library/3.3/xgboost/libs/xgboost.so':
  /home/david/R/x86_64-pc-linux-gnu-library/3.3/xgboost/libs/xgboost.so: undefined symbol: _ZN7xgboost9Predictor4InitERKSt6vectorISt4pairISsSsESaIS3_EERKS1_ISt10shared_ptrINS_7DMatrixEESaISA_EE
Error: loading failed
Execution halted
ERROR: loading failed
* removing ‘/home/david/R/x86_64-pc-linux-gnu-library/3.3/xgboost’
* restoring previous ‘/home/david/R/x86_64-pc-linux-gnu-library/3.3/xgboost’
`

 ldd build/R-package/src/xgboost.so

`
ldd build/R-package/src/xgboost.so
	linux-vdso.so.1 =>  (0x00007ffe3acad000)
	libR.so => /usr/lib/libR.so (0x00007f8ba0758000)
	libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f8ba053a000)
	libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007f8ba0336000)
	librt.so.1 => /lib/x86_64-linux-gnu/librt.so.1 (0x00007f8ba012e000)
	libstdc++.so.6 => /usr/lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007f8b9fda6000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f8b9fa9b000)
	libgomp.so.1 => /usr/lib/x86_64-linux-gnu/libgomp.so.1 (0x00007f8b9f86d000)
	libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007f8b9f656000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f8b9f28f000)
	/lib64/ld-linux-x86-64.so.2 (0x0000564033011000)
	libblas.so.3 => /usr/lib/libblas.so.3 (0x00007f8b9f031000)
	libreadline.so.7 => /lib/x86_64-linux-gnu/libreadline.so.7 (0x00007f8b9ede7000)
	libpcre.so.3 => /lib/x86_64-linux-gnu/libpcre.so.3 (0x00007f8b9eb72000)
	liblzma.so.5 => /lib/x86_64-linux-gnu/liblzma.so.5 (0x00007f8b9e94c000)
	libbz2.so.1.0 => /lib/x86_64-linux-gnu/libbz2.so.1.0 (0x00007f8b9e73c000)
	libz.so.1 => /lib/x86_64-linux-gnu/libz.so.1 (0x00007f8b9e520000)
	libicuuc.so.57 => /usr/lib/x86_64-linux-gnu/libicuuc.so.57 (0x00007f8b9e178000)
	libicui18n.so.57 => /usr/lib/x86_64-linux-gnu/libicui18n.so.57 (0x00007f8b9dcfe000)
	libopenblas.so.0 => /usr/lib/libopenblas.so.0 (0x00007f8b9bbb5000)
	libtinfo.so.5 => /lib/x86_64-linux-gnu/libtinfo.so.5 (0x00007f8b9b98c000)
	libicudata.so.57 => /usr/lib/x86_64-linux-gnu/libicudata.so.57 (0x00007f8b99f0f000)
	libgfortran.so.3 => /usr/lib/x86_64-linux-gnu/libgfortran.so.3 (0x00007f8b99bde000)
	libquadmath.so.0 => /usr/lib/x86_64-linux-gnu/libquadmath.so.0 (0x00007f8b9999e000)
`

**Comment by SteveBronder** [comment]
@davidlamcm can you post your compiler and ubuntu version info as well as your R `sessionInfo()`?

**Comment by davidlamcm** [comment]
```

david@david-linux:~/code/xgboost$ gcc --version
gcc (Ubuntu 4.9.4-2ubuntu1) 4.9.4
Copyright (C) 2015 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

david@david-linux:~/code/xgboost$ g++ --version
g++ (Ubuntu 4.9.4-2ubuntu1) 4.9.4
Copyright (C) 2015 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

david@david-linux:~/code/xgboost$ R

R version 3.3.2 (2016-10-31) -- "Sincere Pumpkin Patch"
Copyright (C) 2016 The R Foundation for Statistical Computing
Platform: x86_64-pc-linux-gnu (64-bit)

> sessionInfo()
R version 3.3.2 (2016-10-31)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 17.04

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_HK.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_HK.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_HK.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_HK.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base  
```

**Comment by khotilov** [comment]
The following builds and installs the xgboost package for me with no issues when running ubuntu 17.04 in docker with CUDA8 installed from ubuntu 16.04 repo https://gitlab.com/nvidia/cuda/blob/ubuntu16.04/8.0/devel/Dockerfile:
```
export CC=/usr/bin/gcc-4.9
export CXX=/usr/bin/g++-4.9
cmake .. -DUSE_CUDA=ON -DR_LIB="ON"
make -j
make install
R
```
```
> library(xgboost)
> sessionInfo()
R version 3.4.2 (2017-09-28)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 17.04

Matrix products: default
BLAS: /usr/lib/libblas/libblas.so.3.7.0
LAPACK: /usr/lib/lapack/liblapack.so.3.7.0

locale:
[1] C

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base

other attached packages:
[1] xgboost_0.6.4.7

loaded via a namespace (and not attached):
[1] compiler_3.4.2      magrittr_1.5        Matrix_1.2-11
[4] tools_3.4.2         stringi_1.1.5       grid_3.4.2
[7] data.table_1.10.4-2 lattice_0.20-35
```
It's necessary to use the same gcc version for everything. 

However, I cannot run the gpu algorithm since I still couldn't figure out the nvidia driver issues with docker. But I might assume that if the driver on your system is properly set up, there should not be any problems with running.

**Comment by davidlamcm** [comment]
@khotilov 
I think you are right. I remove all the files, re-download and re-do all the compile steps again with same version of GCC/G++. And I can install it successfully now. Thanks!

**Comment by SteveBronder** [comment]
Sucess! yes apolgies, I must have been doing something silly
